File: backend/src/services/aiRecommendationService.js
// src/services/aiRecommendationService.js
// Recomendação em texto do Category Management — chamada real à API da
// Claude sobre os números já calculados por categoryAnalyticsService/
// discountThresholdService. Este ficheiro NUNCA consulta a base de dados —
// recebe só os agregados já prontos, para nunca poder inventar um número
// que não veio de lá.
//
// RECUSA-SE A FINGIR (mesmo princípio de multicaixaService.js): sem
// ANTHROPIC_API_KEY configurada, gerar() devolve texto=null com um motivo
// explícito. A análise numérica (volume, threshold, desconto, poupança) não
// depende deste ficheiro e continua a funcionar por inteiro sem IA nenhuma.
const Anthropic = require('@anthropic-ai/sdk');
const config = require('../config/env');
const logger = require('../config/logger');

let client = null;
function getClient() {
  if (!config.anthropic.apiKey) return null;
  if (!client) client = new Anthropic({ apiKey: config.anthropic.apiKey });
  return client;
}

function disponivel() {
  return Boolean(config.anthropic.apiKey);
}

function construirPrompt({ empresa, volumeAtual, categorias, thresholdInfo, oportunidades }) {
  const linhasCategorias = (categorias || [])
    .slice(0, 8)
    .map((c) => `- ${c.categoria}: ${c.valor.toFixed(2)} USD (${c.percentual}% do total)`)
    .join('\n') || 'Sem compras no período.';

  const linhaThreshold = thresholdInfo.proximoThreshold
    ? `Próximo patamar: ${thresholdInfo.proximoThreshold.minVolumeUsd} USD (desconto de ${thresholdInfo.proximoThreshold.discountPercent}%). Faltam ${thresholdInfo.faltamUsd} USD. Poupança adicional potencial ao atingir: ${thresholdInfo.poupancaPotencialUsd} USD.`
    : 'Já está no maior patamar de desconto ativo.';

  const linhasOportunidades = (oportunidades || [])
    .slice(0, 5)
    .map((o) => `- ${o.categoria}: ${o.numeroPos} ordens separadas, faltam ${o.faltamUsd} USD para o desconto de ${o.descontoPotencial}%`)
    .join('\n') || 'Nenhuma identificada no período.';

  return `És um analista de compras B2B. Com base nestes dados REAIS de compra da empresa "${empresa}" nos últimos 12 meses, escreve uma recomendação curta (máximo 120 palavras, em português, tom direto e prático) sobre como aproveitar a economia de escala.

Volume total: ${Number(volumeAtual).toFixed(2)} USD
Desconto atual: ${thresholdInfo.descontoAtual}%
${linhaThreshold}

Categorias com mais volume:
${linhasCategorias}

Oportunidades de consolidação (compras fragmentadas que poderiam cruzar um patamar se juntas):
${linhasOportunidades}

Não inventes números que não estejam aqui. Se não houver nada de acionável, diz isso claramente em vez de forçar uma recomendação.`;
}

/**
 * Gera a recomendação em texto. NUNCA lança por falta de configuração ou
 * falha de rede — devolve { texto: null, motivo } para o resto da análise
 * (que não depende de IA) continuar a responder normalmente.
 */
async function gerar({ empresa, volumeAtual, categorias, thresholdInfo, oportunidades }) {
  const anthropic = getClient();
  if (!anthropic) {
    return { texto: null, motivo: 'ANTHROPIC_API_KEY não está configurada — recomendação de IA desativada.' };
  }

  try {
    const response = await anthropic.messages.create({
      model: config.anthropic.model,
      max_tokens: 400,
      // Tarefa de sumarização/formatação sobre números já calculados, não de
      // raciocínio pesado — effort baixo mantém o custo proporcional a uma
      // chamada que pode acontecer a cada vez que o cliente abre o ecrã.
      output_config: { effort: 'low' },
      messages: [
        { role: 'user', content: construirPrompt({ empresa, volumeAtual, categorias, thresholdInfo, oportunidades }) },
      ],
    });
    const bloco = response.content.find((b) => b.type === 'text');
    const texto = bloco?.text?.trim() || null;
    return { texto, motivo: texto ? null : 'A API não devolveu texto.' };
  } catch (err) {
    logger.warn('aiRecommendationService: falha ao chamar a API da Claude', { error: err.message });
    return { texto: null, motivo: 'Falha ao contactar o serviço de IA — tenta novamente mais tarde.' };
  }
}

module.exports = { disponivel, gerar };

File: backend-java/src/main/java/ao/kixima/analytics/AiRecommendationService.java
package ao.kixima.analytics;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Espelha backend/src/services/aiRecommendationService.js — recomendação
 * curta, em português, a partir dos dados REAIS de compra. Mesmo princípio
 * "recusa-se a fingir": sem ANTHROPIC_API_KEY o texto vem null com o motivo
 * explícito; nunca um texto de IA fabricado. Chamada direta à Messages API
 * (sem SDK), com o mesmo modelo, max_tokens e esforço que o Node.
 */
@Service
public class AiRecommendationService {

    private static final Logger log = LoggerFactory.getLogger(AiRecommendationService.class);

    private final String apiKey;
    private final String model;
    private final ObjectMapper objectMapper;
    private final HttpClient httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(15)).build();

    public AiRecommendationService(@Value("${kixima.anthropic.api-key:}") String apiKey,
                                   @Value("${kixima.anthropic.model:claude-sonnet-5}") String model, ObjectMapper objectMapper) {
        this.apiKey = apiKey == null ? "" : apiKey.trim();
        this.model = model;
        this.objectMapper = objectMapper;
    }

    public boolean disponivel() {
        return !apiKey.isBlank();
    }

    private static String f2(Object v) {
        return String.format(java.util.Locale.ROOT, "%.2f", v == null ? 0d : ((Number) v).doubleValue());
    }

    static String construirPrompt(String empresa, double volumeAtual, List<Map<String, Object>> categorias,
                                  Map<String, Object> thresholdInfo, List<Map<String, Object>> oportunidades) {
        List<String> linhasCategorias = new ArrayList<>();
        for (Map<String, Object> c : categorias.subList(0, Math.min(8, categorias.size()))) {
            linhasCategorias.add("- " + c.get("categoria") + ": " + f2(c.get("valor")) + " USD (" + c.get("percentual") + "% do total)");
        }
        @SuppressWarnings("unchecked")
        Map<String, Object> proximo = (Map<String, Object>) thresholdInfo.get("proximoThreshold");
        String linhaThreshold = proximo != null
                ? "Próximo patamar: " + proximo.get("minVolumeUsd") + " USD (desconto de " + proximo.get("discountPercent") + "%). Faltam "
                + thresholdInfo.get("faltamUsd") + " USD. Poupança adicional potencial ao atingir: " + thresholdInfo.get("poupancaPotencialUsd") + " USD."
                : "Já está no maior patamar de desconto ativo.";
        List<String> linhasOportunidades = new ArrayList<>();
        for (Map<String, Object> o : oportunidades.subList(0, Math.min(5, oportunidades.size()))) {
            linhasOportunidades.add("- " + o.get("categoria") + ": " + o.get("numeroPos") + " ordens separadas, faltam " + o.get("faltamUsd")
                    + " USD para o desconto de " + o.get("descontoPotencial") + "%");
        }
        return "És um analista de compras B2B. Com base nestes dados REAIS de compra da empresa \"" + empresa
                + "\" nos últimos 12 meses, escreve uma recomendação curta (máximo 120 palavras, em português, tom direto e prático) "
                + "sobre como aproveitar a economia de escala.\n\n"
                + "Volume total: " + f2(volumeAtual) + " USD\n"
                + "Desconto atual: " + thresholdInfo.get("descontoAtual") + "%\n"
                + linhaThreshold + "\n\n"
                + "Categorias com mais volume:\n" + (linhasCategorias.isEmpty() ? "Sem compras no período." : String.join("\n", linhasCategorias)) + "\n\n"
                + "Oportunidades de consolidação (compras fragmentadas que poderiam cruzar um patamar se juntas):\n"
                + (linhasOportunidades.isEmpty() ? "Nenhuma identificada no período." : String.join("\n", linhasOportunidades)) + "\n\n"
                + "Não inventes números que não estejam aqui. Se não houver nada de acionável, diz isso claramente em vez de forçar uma recomendação.";
    }

    private static Map<String, Object> resposta(String texto, String motivo) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("texto", texto);
        m.put("motivo", motivo);
        return m;
    }

    /** Nunca lança — devolve `{ texto, motivo }`. */
    public Map<String, Object> gerar(String empresa, double volumeAtual, List<Map<String, Object>> categorias,
                                     Map<String, Object> thresholdInfo, List<Map<String, Object>> oportunidades) {
        if (!disponivel()) return resposta(null, "ANTHROPIC_API_KEY não está configurada — recomendação de IA desativada.");
        try {
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("model", model);
            payload.put("max_tokens", 400);
            payload.put("output_config", Map.of("effort", "low"));
            payload.put("messages", List.of(Map.of("role", "user", "content", construirPrompt(empresa, volumeAtual, categorias, thresholdInfo, oportunidades))));
            HttpRequest req = HttpRequest.newBuilder(URI.create("https://api.anthropic.com/v1/messages"))
                    .timeout(Duration.ofSeconds(60))
                    .header("x-api-key", apiKey)
                    .header("anthropic-version", "2023-06-01")
                    .header("content-type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(payload), StandardCharsets.UTF_8))
                    .build();
            HttpResponse<String> resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString());
            if (resp.statusCode() < 200 || resp.statusCode() >= 300) throw new IllegalStateException("HTTP " + resp.statusCode());
            JsonNode content = objectMapper.readTree(resp.body()).path("content");
            String texto = null;
            for (JsonNode bloco : content) {
                if ("text".equals(bloco.path("type").asText())) {
                    texto = bloco.path("text").asText("").trim();
                    break;
                }
            }
            return resposta(texto == null || texto.isEmpty() ? null : texto, texto == null || texto.isEmpty() ? "A API não devolveu texto." : null);
        } catch (Exception err) {
            log.warn("aiRecommendationService: falha ao chamar a API da Claude: {}", err.getMessage());
            return resposta(null, "Falha ao contactar o serviço de IA — tenta novamente mais tarde.");
        }
    }
}
